### PR TITLE
feat(cli/daemon): boot-time core-relay capability sanity check

### PR DESCRIPTION
**Pull request by branarakic**
## Summary

\`nodeRole: 'core'\` is purely self-declared today. A node can advertise itself as a Core relay and immediately fail at the job because none of its bound interfaces are reachable from the public internet. **beacon-01** in the v10.0.0-rc.10 incident was the canonical case: it bound only to its Tailscale CGNAT address (\`100.99.142.87\`) and could not have functioned as a relay regardless of slot state. The network had no way to detect this — the node just silently failed to serve traffic.

This PR adds a boot-time check that runs after \`agent.start()\` (post libp2p wildcard resolution) and surfaces a structured \`[CORE-PREREQ]\` log if a Core node's actually-bound multiaddrs all classify as non-routable (loopback / RFC1918 / CGNAT / IPv6 ULA / link-local) and no \`announceAddresses\` rescue the verdict.

## Behaviour

| Condition | Log line | Boot continues? |
|---|---|---|
| Core, looks healthy | \`[CORE-PREREQ] OK: N public-class listen addresses bound.\` | Yes |
| Core, looks degraded, \`core.allowDegradedRelay\` omitted/true (**default**) | \`[CORE-PREREQ] WARNING: this Core node looks degraded as a relay. reasons: ...; non-routable addresses: ...\` | Yes (backwards-compatible) |
| Core, looks degraded, \`core.allowDegradedRelay: false\` | \`[CORE-PREREQ] FATAL: ...\` | **No — \`shutdown(1)\`** |
| Edge node, any classification | (no \`[CORE-PREREQ]\` line) | Yes |
| Classifier itself throws | \`[CORE-PREREQ] check failed (continuing boot): ...\` | Yes (defensive) |

**Zero backcompat impact**: every existing operator's behaviour is unchanged unless they explicitly opt into refuse-to-boot semantics.

## Files

| File | Change |
|---|---|
| \`packages/cli/src/daemon/core-prereq-check.ts\` **(new)** | Pure-logic classifier. \`classifyMultiaddr(addr, hostInterfaces)\` + \`checkCoreRelayPrereqs(opts)\`. 11 address classes (\`public\` / \`rfc1918\` / \`cgnat\` / \`loopback\` / \`linkLocal\` / \`ulaIpv6\` / \`dns\` / \`multicast\` / \`wildcardNoPublicInterface\` / \`unknown\` / best-class-across-interfaces for \`/ip4/0.0.0.0\` and \`/ip6/::\`). No I/O — callers inject \`os.networkInterfaces()\` output. |
| \`packages/cli/src/daemon/lifecycle.ts\` (~65 LOC added) | Wires the check after \`await agent.start()\` using \`agent.node.libp2p.getMultiaddrs()\` as the authoritative listenAddress source (post wildcard expansion). Wrapped in try/catch so a classifier crash never blocks the boot. |
| \`packages/cli/src/config.ts\` | New \`core?: { allowDegradedRelay?: boolean }\` field. Future Core-only knobs (e.g. relay-target prioritisation) can live under the same key without polluting the top level. |
| \`packages/cli/src/daemon.ts\` | Barrel re-export. |
| \`packages/cli/test/core-prereq-check.test.ts\` **(new)** | 40 cases — see test plan below. |
| \`packages/cli/vitest.unit.config.ts\` | Added the new test to the unit config so contributors run it via \`pnpm test:unit\` in ~3s instead of paying the ~2-minute hardhat-boot tax. |
| \`CHANGELOG.md\` | Entry under \`[Unreleased]\`. |

## Test plan

\`pnpm --filter @origintrail-official/dkg test:unit test/core-prereq-check.test.ts\` runs the new suite in ~3s:

**Per-class smoke matrix** (\`it.each\`, 21 cases): all 11 classes including the 100.63.x.x / 100.128.x.x edge cases that look like CGNAT but are NOT in 100.64.0.0/10; multicast for v4 and v6; IPv6 prefix patterns (loopback \`::1\`, \`fe80::\` link-local, \`fc00::\`/\`fd00::\` ULA, \`ff::\` multicast, \`2001:db8::\` documentation-as-public).

**Wildcard delegation** (7 cases): \`0.0.0.0\` + public interface; + dual-homed (public + RFC1918, best wins); + only-RFC1918; + only-loopback (\`internal: true\` correctly skipped); + no interfaces; + only-CGNAT (the beacon-01-if-it-had-used-0.0.0.0 scenario); + IPv6 \`::\` analogue.

**7 canonical cases from the plan**:
1. Tailscale-only beacon-01 reproduction (\`/ip4/100.99.142.87/tcp/4001\`) → \`looksDegraded: true\`, reason includes \`1 cgnat\`.
2. \`0.0.0.0\` + public interface → not degraded.
3. \`0.0.0.0\` + only RFC1918 interface → degraded.
4. Single public \`/ip4/203.0.113.5/tcp/4001\` → not degraded.
5. Loopback-only → degraded.
6. IPv6 ULA-only → degraded.
7. DNS-only listen + public \`announceAddresses\` → rescued, not degraded.

**5 safety cases**: edge node skip (\`looksDegraded === false\` for \`'edge'\`); empty \`listenAddresses\` surfaces dedicated reason; mixed-class reasons summarise in a single grep-able line (\`2 cgnat, 1 rfc1918, 1 loopback\`); \`announceAddresses\` without public entries doesn't rescue; missing-\`announceAddresses\` hint surfaces.

- [x] **40/40 tests pass.** \`tsc --noEmit\` clean.

## Out of scope (deliberate split)

- **Pre-start pass** (classify *configured* listenAddresses + interfaces before \`libp2p.start\`). Adds plumbing for ~50ms-earlier warning; the post-start data is authoritative. Future extension.
- **AutoNAT-driven \`natStatus\`** on top of the address classification → PR-5 in the workstream (spike-first; AutoNAT v1 has no public status getter, requires AddressManager event hookup or v2 upgrade).
- **Surface the result in \`/api/status\`** → PR-4 in the workstream; this PR's structured output (\`CorePrereqResult\`) is consumable by that route.
- **\`docs/operator/CORE_RELAY_PREREQS.md\`** runbook referenced in the log message. Doc PR follows independently and will explain how to interpret each non-routable class and what to do about it.

## Related

- **PR-1 #655** (hard shutdown timeout): caught the *symptom* of beacon-01's misconfig (zombie worker on auto-update).
- **PR-2 #659** (\`autoUpdate.source\` flag): removed the *trigger* of beacon-01's incident (git-build path picking up \`main\` HEAD).
- **This PR**: catches the *root condition* (Core node that can't actually serve as a relay) at boot, before the network has to find out the hard way.

Made with [Cursor](https://cursor.com)

### CODE REVIEW

File: packages/cli/src/daemon/core-prereq-check.ts [hunk 1]
```diff
@@ -0,0 +1,309 @@
...
+    .map((c) => c.addr);
+  const nonRoutableAddresses = classified.filter((c) => c.class !== 'public');
+
+  const announcePublic = announceAddresses.some(
+    (a) => classifyMultiaddr(a, hostInterfaces) === 'public',
+  );
+
+  const looksDegraded = nodeRole === 'core'
+    && publicListenAddresses.length === 0
+    && !announcePublic;
```
Comment (branarakic) [inline packages/cli/src/daemon/core-prereq-check.ts]: This lets a public `announceAddresses` entry rescue an otherwise empty bound-address set. If `agent.node.libp2p.getMultiaddrs()` returns `[]` post-start, the core has no local transport bound, so a public announce address cannot make it relay-capable; with `core.allowDegradedRelay: false` this would still boot as OK. Can we keep the degraded verdict when `listenAddresses.length === 0` and only let public announce addresses rescue non-public-but-actually-bound listen addresses?

File: packages/cli/src/daemon/lifecycle.ts [hunk 2]
```diff
@@ -857,6 +859,69 @@ export async function runDaemonInner(
...
+            `non-routable addresses: ${prereq.nonRoutableAddresses
+              .map((a) => `${a.addr} (${a.class})`)
+              .join(', ')}. ` +
+            (allowDegraded
+              ? `Set core.allowDegradedRelay: false in ~/.dkg/config.json to refuse-to-boot on this state. ` +
+                `See docs/operator/CORE_RELAY_PREREQS.md for the full rationale.`
+              : `Refusing to boot. Set core.allowDegradedRelay: true to downgrade this to a warning.`),
+        );
+        if (!allowDegraded) {
+          await shutdown(1);
```
Comment (branarakic) [inline packages/cli/src/daemon/lifecycle.ts]: This fatal path runs before the graceful-shutdown section is initialized. The `shutdown` function declared later closes over `updateInterval`, `chainScanTimer`, `pingTimer`, `pruneTimer`, `rateLimiter`, `metricsCollector`, `server`, etc.; at this point several of those `let`/`const` bindings are still in the temporal dead zone. If a degraded core sets `core.allowDegradedRelay: false`, `await shutdown(1)` can throw (for example `Cannot access updateInterval before initialization`), the surrounding `catch` logs `[CORE-PREREQ] check failed (continuing boot)`, and the node boots even though the operator asked it to refuse. This needs either a pre-initialized minimal shutdown path here (`process.exit(1)` after `agent.stop()` if available) or moving the prereq check after the shutdown cleanup state exists.

File: packages/cli/src/daemon/core-prereq-check.ts [hunk 3]
```diff
@@ -0,0 +1,314 @@
...
+  if (parts.length < 2) return 'unknown';
+  const proto = parts[0];
+  const value = parts[1];
+
+  if (proto === 'dns' || proto === 'dns4' || proto === 'dns6' || proto === 'dnsaddr') {
+    return 'dns';
+  }
+
+  if (proto === 'ip4') {
+    if (value === '0.0.0.0') return bestClassAmongInterfaces(hostInterfaces);
```
[the remark this answers is not in the document]
Comment (branarakic) [reply packages/cli/src/daemon/core-prereq-check.ts]: Addressed in 86ef8a97: wildcard classification is now scoped to the matching address family.

File: packages/cli/src/daemon/core-prereq-check.ts [hunk 4]
```diff
@@ -0,0 +1,314 @@
...
+
+  const publicListenAddresses = classified
+    .filter((c) => c.class === 'public')
+    .map((c) => c.addr);
+  const nonRoutableAddresses = classified.filter((c) => c.class !== 'public');
+
+  const announcePublic = announceAddresses.some(
+    (a) => classifyMultiaddr(a, hostInterfaces) === 'public',
+  );
+  const announceRescues = listenAddresses.length > 0 && announcePublic;
```
[the remark this answers is not in the document]
Comment (branarakic) [reply packages/cli/src/daemon/core-prereq-check.ts]: Addressed in 86ef8a97: public announce rescue no longer clears loopback-only listeners.

File: packages/cli/src/daemon/lifecycle.ts [hunk 5]
```diff
@@ -857,6 +859,79 @@ export async function runDaemonInner(
...
+        const verb = allowDegraded ? 'WARNING' : 'FATAL';
+        log(
+          `[CORE-PREREQ] ${verb}: this Core node looks degraded as a relay. ` +
+            `reasons: ${prereq.reasons.join('; ')}. ` +
+            `non-routable addresses: ${prereq.nonRoutableAddresses
+              .map((a) => `${a.addr} (${a.class})`)
+              .join(', ')}. ` +
+            (allowDegraded
+              ? `Set core.allowDegradedRelay: false in ~/.dkg/config.json to refuse-to-boot on this state. ` +
+                `See docs/operator/CORE_RELAY_PREREQS.md for the full rationale.`
```
[the remark this answers is not in the document]
Comment (branarakic) [reply packages/cli/src/daemon/lifecycle.ts]: Addressed in 86ef8a97: remediation now points at an existing spec document.

File: packages/cli/src/daemon/core-prereq-check.ts [hunk 6]
```diff
@@ -0,0 +1,322 @@
...
+    class: classifyMultiaddr(addr, hostInterfaces),
+  }));
+
+  const publicListenAddresses = classified
+    .filter((c) => c.class === 'public')
+    .map((c) => c.addr);
+  const nonRoutableAddresses = classified.filter((c) => c.class !== 'public');
+
+  const announcePublic = announceAddresses.some(
+    (a) => classifyMultiaddr(a, hostInterfaces) === 'public',
```
[the remark this answers is not in the document]
Comment (branarakic) [reply packages/cli/src/daemon/core-prereq-check.ts]: Addressed in 2b9fbf5d: DNS announce rescue now requires an externally routable hostname instead of treating every DNS multiaddr as public.

File: packages/cli/test/core-prereq-check.test.ts [hunk 7]
```diff
@@ -0,0 +1,314 @@
...
+  mac: '00:00:00:00:00:00',
+  internal: false,
+  cidr: '100.99.142.87/32',
+};
+
+describe('classifyMultiaddr — per-class smoke tests', () => {
+  type Case = [string, AddrClassification];
+  const cases: Case[] = [
+    // The 8 base classes plus wildcards.
+    ['/ip4/203.0.113.5/tcp/4001',     'public'],
```
[the remark this answers is not in the document]
Comment (branarakic) [reply packages/cli/test/core-prereq-check.test.ts]: Addressed in 2b9fbf5d: tests now use real public fixtures and documentation IPv4/IPv6 ranges classify as non-public.

File: packages/cli/src/daemon/core-prereq-check.ts [hunk 8]
```diff
@@ -0,0 +1,324 @@
...
+  }));
+
+  const publicListenAddresses = classified
+    .filter((c) => c.class === 'public')
+    .map((c) => c.addr);
+  const nonRoutableAddresses = classified.filter((c) => c.class !== 'public');
+
+  const announcePublic = announceAddresses.some((a) => {
+    const klass = classifyMultiaddr(a, hostInterfaces);
+    return klass === 'public' || klass === 'dns';
```
[the remark this answers is not in the document]
Comment (branarakic) [reply packages/cli/src/daemon/core-prereq-check.ts]: Addressed in 2b9fbf5d: DNS announce rescue rejects localhost, internal/reserved suffixes, single-label names, and literal-IP hostnames; public DNS coverage uses relay.origintrail.io.

File: packages/cli/src/daemon/lifecycle.ts [hunk 9]
```diff
@@ -857,6 +859,79 @@ export async function runDaemonInner(
 
   await agent.start();
```
[the remark this answers is not in the document]
Comment (branarakic) [reply packages/cli/src/daemon/lifecycle.ts:965]: Addressed in 2b9fbf5d: strict core.allowDegradedRelay=false now runs a pre-start configured-listen classification and exits before binding libp2p sockets when degraded.

File: packages/cli/src/daemon/lifecycle.ts [hunk 10]
```diff
@@ -857,6 +889,77 @@ export async function runDaemonInner(
...
+  // Behaviour gated by `config.core.allowDegradedRelay`:
+  //   - `true` (default): warn-only. No backcompat change for any existing op.
+  //   - `false`: refuse to boot via a minimal post-agent-start cleanup path —
+  //     fail-loud opt-in.
+  //
+  // The post-start pass remains authoritative because libp2p has already
+  // expanded `0.0.0.0` to per-interface bindings here.
+  if (role === 'core') {
+    try {
+      const resolvedMultiaddrs = agent.node.libp2p
```
[the remark this answers is not in the document]
Comment (branarakic) [reply packages/cli/src/daemon/lifecycle.ts]: Addressed in 3e3975c1: the post-start prereq pass now reads transport-manager listener addresses, with getMultiaddrs only as a filtered last-resort fallback. Announce and relay self-addrs no longer feed the direct listener set.

File: packages/cli/src/daemon/core-prereq-check.ts [hunk 11]
```diff
@@ -0,0 +1,377 @@
...
+ * Multiaddr format is `/proto1/value1/proto2/value2/...` — we only need the
+ * first two segments (transport address; the rest is port + circuit/relay
+ * tail). String split keeps us free of a runtime multiaddr-parser
+ * dependency.
+ */
+export function classifyMultiaddr(
+  addr: string,
+  hostInterfaces: ReadonlyArray<NetworkInterfaceInfo> = [],
+): AddrClassification {
+  const parts = addr.split('/').filter(Boolean);
```
[the remark this answers is not in the document]
Comment (branarakic) [reply packages/cli/src/daemon/core-prereq-check.ts:380]: Addressed in 3e3975c1: classifyMultiaddr now detects p2p-circuit in the tail and classifies those addresses as relayed, so relay reservations are excluded from publicListenAddresses.

File: packages/cli/src/daemon/lifecycle.ts [hunk 12]
```diff
@@ -329,6 +331,72 @@ import { handleRequest } from './handle-request.js';
...
+    if (!Array.isArray(listeners)) continue;
+    const addrs = listeners.flatMap((listener) => listenerMultiaddrs(listener));
+    if (addrs.length > 0) return Array.from(new Set(addrs));
+  }
+
+  // Fallback for libp2p shapes that do not expose transport listeners. This
+  // keeps the check operational, while excluding relay reservations so they
+  // cannot masquerade as direct listener capability.
+  if (typeof node.getMultiaddrs === 'function') {
+    return stringifyMultiaddrList(node.getMultiaddrs())
```
[the remark this answers is not in the document]
Comment (branarakic) [reply packages/cli/src/daemon/lifecycle.ts]: Addressed in 9c5c50b1: removed the getMultiaddrs fallback from listener evidence, so libp2p shapes without transport listeners now produce an indeterminate/degraded result instead of treating advertised addresses as bound listeners.

File: packages/cli/src/daemon/core-prereq-check.ts [hunk 13]
```diff
@@ -0,0 +1,382 @@
...
+    class: classifyMultiaddr(addr, hostInterfaces),
+  }));
+
+  const publicListenAddresses = classified
+    .filter((c) => c.class === 'public')
+    .map((c) => c.addr);
+  const nonRoutableAddresses = classified.filter((c) => c.class !== 'public');
+
+  const announcePublic = announceAddresses.some((a) => isPublicAnnounceAddress(a, hostInterfaces));
+  const announceCanServe = classified.some(
```
[the remark this answers is not in the document]
Comment (branarakic) [reply packages/cli/src/daemon/core-prereq-check.ts:426]: Addressed in 9c5c50b1: wildcardNoPublicInterface now participates in the announce rescue path, so synthetic pre-start wildcard listeners can be rescued by a public announce address while still reporting the underlying wildcard class.

File: packages/cli/src/daemon/lifecycle.ts [hunk 14]
```diff
@@ -329,6 +331,68 @@ import { handleRequest } from './handle-request.js';
...
+  for (const manager of managers) {
+    const getListeners = (manager as { getListeners?: () => unknown } | undefined)?.getListeners;
+    if (typeof getListeners !== 'function') continue;
+    const listeners = getListeners.call(manager);
+    if (!Array.isArray(listeners)) continue;
+    const addrs = listeners.flatMap((listener) => listenerMultiaddrs(listener));
+    if (addrs.length > 0) return Array.from(new Set(addrs));
+  }
+
+  // If this libp2p shape does not expose transport listeners, prefer an
```
[the remark this answers is not in the document]
Comment (branarakic) [reply packages/cli/src/daemon/lifecycle.ts:397]: Addressed in fd6482ff: a listener introspection miss now returns an indeterminate result and skips the post-start relay prerequisite verdict instead of feeding an empty listen set into the fatal/degraded path.

File: packages/cli/src/daemon/core-prereq-check.ts [hunk 15]
```diff
@@ -0,0 +1,383 @@
...
+    (c) => c.class === 'dns'
+      || c.class === 'rfc1918'
+      || c.class === 'cgnat'
```
[the remark this answers is not in the document]
Comment (branarakic) [reply packages/cli/src/daemon/core-prereq-check.ts]: Fixed in c6abcc1f: public announce rescue is now limited to RFC1918 and unresolved-wildcard listeners only. Explicit CGNAT and ULA listeners remain degraded even with a literal public announce address, with regression coverage for both cases.

File: packages/cli/src/daemon/core-prereq-check.ts [hunk 16]
```diff
@@ -0,0 +1,383 @@
...
+function isPublicDnsHostname(hostRaw: string | undefined): boolean {
+  if (!hostRaw) return false;
+  const host = hostRaw.toLowerCase().replace(/\.$/, '');
+  if (!host || host === 'localhost') return false;
+  const ipFamily = isIP(host);
+  if (ipFamily === 4) return classifyIPv4(host) === 'public';
+  if (ipFamily === 6) return classifyIPv6(host) === 'public';
+  if (!host.includes('.')) return false;
+  if (!/^[a-z0-9-]+(?:\.[a-z0-9-]+)+$/.test(host)) return false;
+  return !(
```
[the remark this answers is not in the document]
Comment (branarakic) [reply packages/cli/src/daemon/core-prereq-check.ts]: Fixed in c6abcc1f: unresolved DNS announce addresses no longer count as public evidence. The pure checker now only trusts literal public IP announce addresses, so typo or nonexistent FQDNs cannot bypass strict degraded verdicts.

File: packages/cli/src/daemon/core-prereq-check.ts [hunk 17]
```diff
@@ -0,0 +1,348 @@
...
+    addr,
+    class: classifyMultiaddr(addr, hostInterfaces),
+  }));
+
+  const publicListenAddresses = classified
+    .filter((c) => c.class === 'public')
+    .map((c) => c.addr);
+  const nonRoutableAddresses = classified.filter((c) => c.class !== 'public');
+
+  const announcePublic = announceAddresses.some((a) => isPublicAnnounceAddress(a, hostInterfaces));
```
[the remark this answers is not in the document]
Comment (branarakic) [reply packages/cli/src/daemon/core-prereq-check.ts:425]: Addressed in 58272496: DNS announce now triggers a soft (indeterminate) rescue rather than a hard refusal — `looksDegraded: false, indeterminate: true` so strict `allowDegradedRelay: false` no longer refuses to boot stable-DNS deployments. Reserved DNS names (`localhost`, `*.local`, `*.test`, `*.example`, `*.invalid`, single-label) are excluded from the rescue. The lifecycle layer can pre-resolve DNS and re-call the helper for hard validation.

File: packages/cli/src/daemon/core-prereq-check.ts [hunk 18]
```diff
@@ -0,0 +1,348 @@
...
+
+function classifyIPv4(ip: string): AddrClassification {
+  if (!isValidIPv4(ip)) return 'unknown';
+  if (isLoopbackIPv4(ip)) return 'loopback';
+  if (isLinkLocalIPv4(ip)) return 'linkLocal';
+  if (isCgnatIPv4(ip)) return 'cgnat';
+  if (isRfc1918IPv4(ip)) return 'rfc1918';
+  if (isMulticastIPv4(ip)) return 'multicast';
+  if (isDocumentationIPv4(ip)) return 'unknown';
+  return 'public';
```
[the remark this answers is not in the document]
Comment (branarakic) [reply packages/cli/src/daemon/core-prereq-check.ts:242]: Addressed in 58272496: switched IPv4 classification from a small "looks public" allow-list to "reject if reserved, else public". Now explicitly classifies `0.0.0.0/8` (RFC 1122 "this network"), `198.18.0.0/15` (RFC 2544 benchmark testing), and `240.0.0.0/4` (RFC 1112 reserved-future-use) as a new `'reserved'` class so they show up in `nonRoutableAddresses` instead of slipping through as `public`. Three regression tests cover the three blocks the comment called out (`198.18.1.1`, `0.1.2.3`, `240.0.0.1`).

File: packages/cli/src/daemon/core-prereq-check.ts [hunk 19]
```diff
@@ -0,0 +1,348 @@
...
+  if (/^f[cd][0-9a-f]{2}:/.test(ip)) return 'ulaIpv6';
+  if (/^ff[0-9a-f]{2}:/.test(ip)) return 'multicast';
+  return 'public';
+}
+
+function isPublicAnnounceAddress(
+  addr: string,
+  hostInterfaces: ReadonlyArray<NetworkInterfaceInfo>,
+): boolean {
+  const klass = classifyMultiaddr(addr, hostInterfaces);
```
[the remark this answers is not in the document]
Comment (branarakic) [reply packages/cli/src/daemon/core-prereq-check.ts:291]: Addressed in 58272496: `isPublicAnnounceAddress()` now explicitly rejects `/ip4/0.0.0.0` and `/ip6/::` via a new `isWildcardMultiaddr()` guard before delegating to `classifyMultiaddr()`, so a wildcard announce can no longer suppress the degraded verdict by riding on a public host interface. Two regression tests cover both families plus a positive test verifying that mixing a wildcard with a literal-public announce still rescues via the literal-public entry.
